drivers/platform/aducm3029/irq.c

```diff
@@ -0,0 +1,298 @@
+
+/******************************************************************************/
+/************************* Include Files **************************************/
+/******************************************************************************/
+
+#include "irq.h"
+#include "irq_extra.h"
+#include "error.h"
+#include <stdlib.h>
+
+/******************************************************************************/
+/********************** Macros and Constants Definitions **********************/
+/******************************************************************************/
+
+/** The number of the first external interrupt, used by NVIC */
+#define BASE_XINT_NB			(XINT_EVT0_IRQn)
+
+/** Number of interrupts controllers available */
+#define NB_INTERRUPT_CONTROLLERS	1u
+
+/** Map the interrupt ID to the ADI_XINT_EVENT associated event */
+static const uint32_t id_map_event[NB_EXT_INTERRUPTS] = {
+	ADI_XINT_EVENT_INT0, // ID 0
+	ADI_XINT_EVENT_INT1, // ID 1
+	ADI_XINT_EVENT_INT2, // ID 2
+	ADI_XINT_EVENT_INT3  // ID 3
+};
+
+/******************************************************************************/
+/***************************** Global Variables *******************************/
+/******************************************************************************/
+
+/** Bitmap with initialized interrupt */
+static uint32_t		initialized;
+
+/******************************************************************************/
+/************************ Functions Definitions *******************************/
+/******************************************************************************/
+
+/**
+ * @brief Call the user callback
+ * @param aducm_desc - Descriptor where the user callback is stared
+ * @param event - Event that generated the callback
+ * @param arg - Unused
+ */
+static void internal_callback(void *aducm_desc, uint32_t event, void *arg)
+{
+	struct aducm_irq_desc *desc = aducm_desc;
+
+	(void)arg;
+
+	switch (event) {
+	case ADI_XINT_EVENT_INT0:
+		if (desc->irq_handler[0])
+			desc->irq_handler[0]((void *)0);
+		break;
+	case ADI_XINT_EVENT_INT1:
+		if (desc->irq_handler[1])
+			desc->irq_handler[1]((void *)1);
+		break;
+	case ADI_XINT_EVENT_INT2:
+		if (desc->irq_handler[2])
+			desc->irq_handler[2]((void *)2);
+		break;
+	case ADI_XINT_EVENT_INT3:
+		if (desc->irq_handler[3])
+			desc->irq_handler[3]((void *)3);
+		break;
+	default:
+		break;
+	}
+}
+
+/**
+ * @brief Initialized the controller for the ADuCM3029 external interrupts
+ *
+ * @param desc - Pointer where the configured instance is stored
+ * @param param - Configuration information for the instance
+ * @return \ref SUCCESS in case of success, \ref FAILURE otherwise.
+ */
+int32_t irq_ctrl_init(struct irq_ctrl_desc **desc,
+		      const struct irq_init_param *param)
+{
+	struct aducm_irq_desc		*aducm_desc;
+
+	if (!desc || !param || initialized)
+		return FAILURE;
+
+	if (!(*desc = calloc(1, sizeof(**desc))))
```

Comment:
It may be possible for the compiler to interpret this statement differently, depending on the used c language standard.

Reply:
solved